src/ds/ds6/hw_monitor_extended_buffers.cpp

```diff
@@ -36,80 +27,90 @@ namespace librealsense
-                newCommand.data,
-                newCommand.sizeOfSendCommandData,
-                details.sendCommandData.data(),
-                details.sizeOfSendCommandData);
-
-            if (locked_transfer)
-            {
-                return _locked_transfer->send_receive({ details.sendCommandData.begin(),details.sendCommandData.end() });
-            }
-
-            send_hw_monitor_command(details);
-
-            // Error/exit conditions
-            if (p_response)
-                *p_response = hwm_Success;
-            if (newCommand.oneDirection)
-                return std::vector<uint8_t>();
-
-            librealsense::copy(newCommand.receivedOpcode, details.receivedOpcode.data(), 4);
-            librealsense::copy(newCommand.receivedCommandData, details.receivedCommandData.data(), details.receivedCommandDataLength);
-            newCommand.receivedCommandDataLength = details.receivedCommandDataLength;
-
-            // endian?
-            auto opCodeAsUint32 = pack(details.receivedOpcode[3], details.receivedOpcode[2],
-                details.receivedOpcode[1], details.receivedOpcode[0]);
-            if (opCodeAsUint32 != opCodeXmit)
-            {
-                auto err_type = static_cast<hwmon_response>(opCodeAsUint32);
-                std::string err = hwmon_error_string(cmd, err_type);
-                LOG_DEBUG(err);
-                if (p_response)
-                {
-                    *p_response = err_type;
-                    return std::vector<uint8_t>();
-                }
-                throw invalid_value_exception(err);
-            }
-            auto ans = std::vector<uint8_t>(newCommand.receivedCommandData,
-                newCommand.receivedCommandData + newCommand.receivedCommandDataLength);
+            // preparing data to be sent in current chunk
+            cmd.data = get_data_for_current_iteration(table_data, i);
+            // chunk number is in param4
+            cmd.param4 = compute_chunks_param(overall_chunks, i);
 
-            recv_msg.insert(recv_msg.end(), ans.begin(), ans.end());
+            answer = hw_monitor::send(cmd, p_response, locked_transfer);
         }
+        return answer;
 
-        return recv_msg;
     }
+
+    std::vector<uint8_t> hw_monitor_extended_buffers::get_data_for_current_iteration(const std::vector<uint8_t>& table_data, int iteration) const
+    {
+        auto remaining_data_size_to_be_sent = table_data.size() - iteration * HW_MONITOR_COMMAND_SIZE;
+
+        auto start_iterator_for_current_chunk = table_data.begin() + iteration * HW_MONITOR_COMMAND_SIZE;
+        auto end_iterator_for_current_chunk = (remaining_data_size_to_be_sent >= HW_MONITOR_COMMAND_SIZE) ?
+            start_iterator_for_current_chunk + HW_MONITOR_COMMAND_SIZE :
+            start_iterator_for_current_chunk + remaining_data_size_to_be_sent;
+
+        return std::vector<uint8_t>(start_iterator_for_current_chunk, end_iterator_for_current_chunk);
```

Comment:
Consider getting the `cmd.data` as a & parame to the function (or returning with std::move?).
Here you allocate a new vector , copy the data to it and then copy the return vector to cmd.data.

Reply:
Added std::move